### PR TITLE
feat: smarter handling of references and versions in dependency configs [APE-829]

### PR DESCRIPTION
**Pull request by unparalleled-js**
### What I did

fixes: #1381 
Fixes: APE-820

### How I did it

* document stuff about tags
* if version not found, try as a reference
* if missing `v` prefix, try with `v` prefix.

### How to verify it

this will work now:

```
dependencies:
  - name: Uniswap
    github: Uniswap/v3-core
    version: 1.0.0
```

tho youll get a warning.

best to do this:

```

dependencies:
  - name: Uniswap
    github: Uniswap/v3-core
    ref: v1.0.0
```

### Checklist

- [ ] All changes are completed
- [ ] New test cases have been added
- [ ] Documentation has been updated

### CODE REVIEW

File: docs/userguides/config.md [hunk 1]
```diff
@@ -45,6 +45,19 @@ dependencies:
...
+For example, to install a version based on a `git` tag, do the following:
+
+```yaml
+dependencies:
+  - name: Uniswap
+    github: Uniswap/v3-core
+    branch: v1.0.0
+```
+
+The `branch:` config installs the code from that branch or tag whereas the `version:` config uses the official GitHub release API.
```
Comment (fubuloubu) [inline docs/userguides/config.md:59]: Wonder if `branch` is the confusing bit and it should be `ref` or something (breaking change obvs)

**Review by fubuloubu — APPROVED** [review]

File: docs/userguides/config.md (hunk 1 above)
Comment (unparalleled-js) [reply docs/userguides/config.md:59]: We could deprecate `branch:`, add `ref:`, make `branch:` be an alias and undocument it, and then remove `branch:` on 0.7. We could do that now.

**Comment by unparalleled-js** [comment]
another idea that could alongside: if the version release is not found from `version:`, check the tags. maybe warn as well saying to use `branch:` (or `ref:`), but then it at least would still work.

thoughts there?

**Comment by fubuloubu** [comment]
> another idea that could alongside: if the version release is not found from `version:`, check the tags. maybe warn as well saying to use `branch:` (or `ref:`), but then it at least would still work.
> 
> thoughts there?

Both the suggestions are good suggestions, I think `ref:` is clearer to the type of git operation that's being performed (vs. `branch:` which is just one type of `ref:`)

File: src/ape/managers/project/dependency.py [hunk 2]
```diff
@@ -61,19 +62,36 @@ class GithubDependency(DependencyAPI):
     such as ``dapphub/erc20``.
     """
 
+    # TODO: Remove at >= 0.7
```
Comment (unparalleled-js) [inline src/ape/managers/project/dependency.py:65]: APE-830, it is in the 0.7 view

**Review by fubuloubu — APPROVED** [review]
This seems more robust!

**Pull request merged**